### PR TITLE
Add B2C integration tests.

### PR DESCRIPTION
**Pull request by ZhuXiaoBing-cn**
Add B2C integration tests.

### CODE REVIEW

File: sdk/spring/azure-spring-boot-test-b2c/pom.xml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  <parent>
+    <groupId>org.springframework.boot</groupId>
+    <artifactId>spring-boot-starter-parent</artifactId>
+    <version>2.3.7.RELEASE</version> <!-- {x-version-update;org.springframework.boot:spring-boot-starter-parent;external_dependency} -->
+    <relativePath/> <!-- lookup parent from repository -->
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.azure.spring</groupId>
+  <artifactId>azure-spring-boot-test-b2c</artifactId>
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-test-b2c/pom.xml:14]: `b2c` -> `aad-b2c`. Same to other places like: folder name, environment variable name.

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    public B2CSeleniumITHelper(Class<?> appClass, Map<String, String> properties) throws InterruptedException {
+        emailAddress = B2C_USER_EMAIL;
+        password =  System.getenv(B2C_USER_PASSWORD);
+        app = new AppRunner(appClass);
+        DEFAULT_PROPERTIES.forEach(app::property);
+        properties.forEach(app::property);
+
+        ChromeOptions options = new ChromeOptions();
+//        options.addArguments("--headless");
+//        options.addArguments("--incognito", "--no-sandbox", "--disable-dev-shm-usage");
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java:81]: The `//` added for debug purpose?

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/utils/B2CTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+    public static final String B2C_USER_EMAIL = "chrisgreen@abc.com";
+    public static final String B2C_USER_PASSWORD = "B2C_USER_PASSWORD";
+    public static final String B2C_TENANT = "B2CITTest1";
+    public static final String B2C_CLIENT_ID = "B2C_CLIENT_ID";
+    public static final String B2C_CLIENT_SECRET = "B2C_CLIENT_SECRET";
+    public static final String B2C_REPLY_URL = "http://localhost:8080/home";
+    public static final String B2C_SIGN_UP_OR_SIGN_IN = "B2C_1_signupsignin1";
+    public static final String B2C_PROFILE_EDIT = "B2C_1_profileediting1";
+}
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/utils/B2CTestUtils.java:13]: Please remove duplicated files of `chromedeiver_xxx`.

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/utils/B2CTestUtils.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+public class B2CTestUtils {
+
+    public static final String B2C_USER_EMAIL = "chrisgreen@abc.com";
+    public static final String B2C_USER_PASSWORD = "B2C_USER_PASSWORD";
+    public static final String B2C_TENANT = "B2CITTest1";
+    public static final String B2C_CLIENT_ID = "B2C_CLIENT_ID";
+    public static final String B2C_CLIENT_SECRET = "B2C_CLIENT_SECRET";
+    public static final String B2C_REPLY_URL = "http://localhost:8080/home";
+    public static final String B2C_SIGN_UP_OR_SIGN_IN = "B2C_1_signupsignin1";
+    public static final String B2C_PROFILE_EDIT = "B2C_1_profileediting1";
```
Comment (chenrujun) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/utils/B2CTestUtils.java:12]: Add `AAD_` prefix.

File: sdk/spring/tests.yml [hunk 5]
```diff
@@ -52,6 +52,9 @@ jobs:
           AAD_SINGLE_TENANT_CLIENT_SECRET: $(java-spring-aad-single-tenant-secret)
           AAD_SINGLE_TENANT_CLIENT_ID_WITH_ROLE: $(java-spring-aad-single-tenant-id-with-role)
           AAD_SINGLE_TENANT_CLIENT_SECRET_WITH_ROLE: $(java-spring-aad-single-tenant-secret-with-role)
+          B2C_USER_PASSWORD: $(java-spring-b2c-user-password)
+          B2C_CLIENT_ID: $(java-spring-b2c-client-id)
+          B2C_CLIENT_SECRET: $(java-spring-b2c-client-secret)
```
Comment (chenrujun) [inline sdk/spring/tests.yml:63]: Add `AAD_` prefix.

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/AADB2CIT.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+
+public class AADB2CIT {
+
+    private final String A_WORKER = "a worker";
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/AADB2CIT.java:26]: How about add a prefix `JOB_TITLE_` to these fields?

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/AADB2CIT.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        Assert.assertNotNull(jobTitle);
+        Assert.assertEquals(newJobTitle, jobTitle);
+        Assert.assertEquals(B2C_PROFILE_EDIT, userFlowName);
+    }
+
+    @Test
+    public void testLogOut() throws InterruptedException {
+        B2CSeleniumITHelper b2CSeleniumITHelper = new B2CSeleniumITHelper(DumbApp.class, Collections.emptyMap());
+        b2CSeleniumITHelper.signIn(B2C_SIGN_UP_OR_SIGN_IN);
+        b2CSeleniumITHelper.logout();
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/AADB2CIT.java:62]: Do we need any assertions here?

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java (hunk 2 above)
Comment (saragluna) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java:81]: What are these two lines for?

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            throw new RuntimeException(e);
+        } finally {
+            if (process != null) {
+                process.destroy();
+            }
+        }
+    }
+
+    public B2CSeleniumITHelper(Class<?> appClass, Map<String, String> properties) throws InterruptedException {
+        emailAddress = B2C_USER_EMAIL;
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java:73]: how about change `emailAddress` to `userEmail` and `password` to `userPassword` to make them consistent.

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        Thread.sleep(5000);
+        driver.findElement(By.id("logout")).click();
+        Thread.sleep(3000);
+        driver.findElement(By.cssSelector("button[type='submit']")).click();
+        Thread.sleep(3000);
+        manualRedirection();
+        driver.findElement(By.cssSelector("a[href='/oauth2/authorization/" + B2C_SIGN_UP_OR_SIGN_IN + "']")).click();
+        Thread.sleep(3000);
+        String name = driver.findElement(By.cssSelector("button[type='submit']")).getText();
+        Assert.assertEquals("Sign in",name);
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java:120]: The assertion should be moved to the IT class, the selenium should only be responsible for interaction with the browser.

Comment (saragluna) [reply sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/selenium/B2CSeleniumITHelper.java:120]: We could provide a function called 
```
public getLoginButtonText() {
}
```

and do the verification in your test.

File: sdk/spring/tests.yml [hunk 10]
```diff
@@ -52,6 +52,9 @@ jobs:
           AAD_SINGLE_TENANT_CLIENT_SECRET: $(java-spring-aad-single-tenant-secret)
           AAD_SINGLE_TENANT_CLIENT_ID_WITH_ROLE: $(java-spring-aad-single-tenant-id-with-role)
           AAD_SINGLE_TENANT_CLIENT_SECRET_WITH_ROLE: $(java-spring-aad-single-tenant-secret-with-role)
+          B2C_USER_PASSWORD: $(java-spring-b2c-user-password)
```
Comment (saragluna) [inline sdk/spring/tests.yml:61]: Please also add them in the below task's env section:
```yaml
 - task: Maven@3
   displayName: 'Run tests again, JDK Version: $(JavaBuildVersion)'
```

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/utils/B2CTestUtils.java (hunk 3 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot-test-b2c/src/test/java/com/azure/test/b2c/utils/B2CTestUtils.java:13]: This will be removed in the next PR.Use a public `chromedeiver_xxx`.

File: sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/utils/AADB2CTestUtils.java [hunk 11]
```diff
@@ -2,13 +2,13 @@
...
-    public static final String AAD_B2C_SIGN_UP_OR_SIGN_IN = "AAD_B2C_SIGN_UP_OR_SIGN_IN";
-    public static final String AAD_B2C_PROFILE_EDIT = "AAD_B2C_PROFILE_EDIT";
+    public static final String AAD_B2C_USER_EMAIL =  System.getenv("AAD_B2C_USER_EMAIL");
+    public static final String AAD_B2C_USER_PASSWORD = System.getenv("AAD_B2C_USER_PASSWORD");
+    public static final String AAD_B2C_TENANT = System.getenv("AAD_B2C_TENANT");
+    public static final String AAD_B2C_CLIENT_ID = System.getenv("AAD_B2C_CLIENT_ID");
+    public static final String AAD_B2C_CLIENT_SECRET = System.getenv("AAD_B2C_CLIENT_SECRET");
+    public static final String AAD_B2C_REPLY_URL = System.getenv("AAD_B2C_REPLY_URL");
+    public static final String AAD_B2C_SIGN_UP_OR_SIGN_IN = System.getenv("AAD_B2C_SIGN_UP_OR_SIGN_IN");
+    public static final String AAD_B2C_PROFILE_EDIT = System.getenv("AAD_B2C_PROFILE_EDIT");
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/utils/AADB2CTestUtils.java:12]: How about changing the class name to AADB2CEnvrionemt? We could do it in another PR.

Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/utils/AADB2CTestUtils.java:12]: Okay

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        Assert.assertEquals(AADB2CTestUtils.AAD_B2C_PROFILE_EDIT, userFlowName);
+    }
+
+    @Test
+    public void testLogOut() throws InterruptedException {
+        AADB2CSeleniumITHelper AADB2CSeleniumITHelper = new AADB2CSeleniumITHelper(DumbApp.class,
+            Collections.emptyMap());
+        AADB2CSeleniumITHelper.signIn(AADB2CTestUtils.AAD_B2C_SIGN_UP_OR_SIGN_IN);
+        AADB2CSeleniumITHelper.logout();
+        String signInButtonText = AADB2CSeleniumITHelper.getSignInButtonText();
```
[the remark this answers is not in the document]
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java:64]: Okay.

File: sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        Assert.assertNotNull(jobTitle);
+        Assert.assertEquals(newJobTitle, jobTitle);
+        Assert.assertEquals(AADB2CTestUtils.AAD_B2C_PROFILE_EDIT, userFlowName);
+    }
+
+    @Test
+    public void testLogOut() throws InterruptedException {
+        AADB2CSeleniumITHelper AADB2CSeleniumITHelper = new AADB2CSeleniumITHelper(DumbApp.class,
+            Collections.emptyMap());
+        AADB2CSeleniumITHelper.signIn(AADB2CTestUtils.AAD_B2C_SIGN_UP_OR_SIGN_IN);
```
[the remark this answers is not in the document]
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java:62]: Okay.

**Review by saragluna — APPROVED** [review]
LGTM.

File: sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (saragluna) [reply sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java]: I don't think we should combine these into one method. Logout is an independent operation, while getSignInButtonText is only to retrieve some result, but it doesn't have to be the result of logout. I prefer to keep them separate.

File: sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (saragluna) [reply sdk/spring/azure-spring-boot-test-aad-b2c/src/test/java/com/azure/test/aad/b2c/selenium/AADB2CIT.java]: I would prefer keeping AADB2CSeleniumITHelper's constructor simple, without taking any signin action here.
